pytorch_lightning/trainer/supporters.py

```diff
@@ -120,23 +130,126 @@ def mean(self):
+        internal_predictions = self.predictions
+
+        if dl_idx not in internal_predictions:
+            internal_predictions[dl_idx] = {}
+
+        for pred in predictions:
+            if "path" in pred:
+                if self.should_all_gather:
+                    raise MisconfigurationException(
+                        "`path` key is not supported yet in distributed setting. Please, use `id`. "
+                    )
+                key = pred["path"]
+
+            elif "id" in pred:
+                key = pred["id"]
+                if not isinstance(key, (int, float, torch.Tensor)):
+                    raise MisconfigurationException(
+                        f"`id` key should be either a int or float. Found {type(key)}."
+                    )
+
+            else:
+                raise MisconfigurationException(
+                    "When predictions are provided within a dict, we expect either a `path` or `id` key. "
+                )
+
+            if isinstance(key, torch.Tensor):
+                key = str(key.item())
+
+            if key not in internal_predictions[dl_idx]:
+                internal_predictions[dl_idx][key] = []
+            else:
+                raise MisconfigurationException(
+                    "Prediction Collection doesn't support multiple prediction for one sample yet.")
+
+            internal_predictions[dl_idx][key] = pred
+
+    def cache(self, predictions: List, dl_idx: int, current_stage):
+        if predictions is None:
+            return
+
+        self.current_stage = current_stage
+
+        assert isinstance(predictions, (list, tuple))
+        if not all([isinstance(p, (dict)) for p in predictions]):
+            raise MisconfigurationException(
+                "predictions objects should be a list or tuple. "
+                "Each contained element should be either a dict. "
+            )
+
+        if not all([len(p) > 1 for p in predictions]):
+            raise MisconfigurationException(
+                "predictions objects should be a list or tuple. "
+                "Each contained element should contain at minimum an ID and a prediction tensor. "
+            )
+
+        self._cache_prediction(predictions, dl_idx)
+
+    def attach_predictions(self, results, current_stage: str) -> list:
+        self.current_stage = current_stage
+        if len(self) > 0:
+            predictions = self.predictions
+            for dl_idx, result in enumerate(results):
+                if dl_idx in predictions:
+                    dl_predictions = predictions[dl_idx]
+                    dl_predictions = self.allgather_predictions(dl_predictions)
+                    result["predictions"] = [*dl_predictions.values()]
+        return results
+
+    def allgather_predictions(self, predictions):
+        """
+        This function all_gather predictions accross multiple processes
+        #todo: Implement a more stable version
```

Comment:
Can you improve this TODO so a future visitor knows what makes this version unstable?

Reply:
lest create already an issue for it and link it here